### ISSUE TITLE
 Test failure: test_simulate_route (bark_spider.tests.SimulateRouteTests)

### ISSUE THREAD

**Issue by rob-smallshire** [sixty-north/bark-spider#19]
```
$ python -m unittest bark_spider.tests
F
======================================================================
FAIL: test_simulate_route (bark_spider.tests.SimulateRouteTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/rjs/dev/bark-spider/bark_spider/tests.py", line 108, in test_simulate_route
    expected)
AssertionError: {'ela[19 chars]3', '0': '0', '7': '7', '2': '2', '4': '4', '1[477 chars]'1'}} != {'ela[19 chars]3', '9': '9', '0': '0', '7': '7', '4': '4', '1[478 chars]'1'}}
Diff is 1637 characters long. Set self.maxDiff to None to see it.

----------------------------------------------------------------------
Ran 1 test in 0.782s

FAILED (failures=1)
```

I haven't investigated further.

**Comment by abingham** [comment]
This is almost certainly a dict-key ordering issue. The test doesn't make any attempt to sort the keys as it should (or perhaps there's a dict-equals assertion we can use).

**Comment by abingham** [comment]
I haven't been able to replicate this yet. Can you run it with `self.maxDiff = None` (just insert that code at the beginning of `bark_spider.tests.SimulateRouteTests.test_simulate_route`)? My initial assessment was probably wrong, and it looks like we're getting different simulation results on different systems.

**Comment by rob-smallshire** [comment]
Weirdly, if I run the tests (unchanged) from inside PyCharm they pass.

**Comment by rob-smallshire** [comment]
Running from the terminal with maxDiff = None:

```
$ python -m unittest bark_spider.tests
F
======================================================================
FAIL: test_simulate_route (bark_spider.tests.SimulateRouteTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/rjs/dev/bark-spider/bark_spider/tests.py", line 110, in test_simulate_route
    expected)
AssertionError: {'ela[41 chars]6', '3': '3', '9': '9', '2': '2', '5': '5', '7[455 chars]'4'}} != {'ela[41 chars]6', '5': '5', '4': '4', '3': '3', '7': '7', '2[456 chars]'1'}}
  {'elapsed_time': {'0': '0',
                    '1': '1',
                    '10': '10',
                    '2': '2',
                    '3': '3',
                    '4': '4',
                    '5': '5',
                    '6': '6',
                    '7': '7',
                    '8': '8',
                    '9': '9'},
   'software_development_rate': {'0': '0',
-                                '1': '1.1458134842011096',
+                                '1': '0.8065923869047285',
                                 '10': '0',
-                                '2': '1.088522809991054',
-                                '3': '1.0340966694915015',
-                                '4': '0.9823918360169261',
?                                 ^      -  --   ^^^^^ ^

+                                '2': '0.8567217268385271',
?                                 ^       +++  +++  ^^ ^

+                                '3': '0.9043445997756356',
+                                '4': '0.9495863290658887',
                                 '5': '0.9925659718916292',
                                 '6': '1.0333966325760826',
                                 '7': '1.0721857602263134',
                                 '8': '1.109035431494033',
                                 '9': '1.1440426191983664'},
   'step_number': {'0': '0',
                   '1': '1',
                   '10': '10',
                   '2': '2',
                   '3': '3',
                   '4': '4',
                   '5': '5',
                   '6': '6',
                   '7': '7',
                   '8': '8',
                   '9': '9'}}

----------------------------------------------------------------------
Ran 1 test in 0.719s

FAILED (failures=1)
```

**Comment by rob-smallshire** [comment]
Looking at the data structures displayed above, is it intentional that all these integers and floats are being represented as strings?

**Comment by abingham** [comment]
Yes, insofar as that's what Pyramid is sending to the client by default.

**Comment by rob-smallshire** [comment]
For what it's worth, my results are consistent.  The numbers aren't changing between runs.

**Comment by rob-smallshire** [comment]
Also Python version is as follows:

$ python
Python 3.4.3 (v3.4.3:9b73f1c3e601, Feb 23 2015, 02:52:03) 
[GCC 4.2.1 (Apple Inc. build 5666) (dot 3)] on darwin

**Comment by abingham** [comment]
OK, I'll try it out this evening on a mac as well. Curiouser and curiouser...

**Comment by rob-smallshire** [comment]
I'm retracting – and have deleted – my earlier remark about the tests passing in PyCharm. I was running the other the tests.  I now get exactly the same inside and outside of PyCharm.

**Comment by abingham** [comment]
I have now been able to replicate this on os x. My guess is that this may just be a difference in how brooks behaves. Ultimately that's what we're testing here, so if brooks changed between the time I captured the expected results and now, then we just need to update the expected results.

What I'll do is update brooks tomorrow on a machine where the tests are passing. If that breaks things, then I think we've clinched it.

**Comment by rob-smallshire** [comment]
If that's the case, this test doesn't belong in bark spider in its current form. Can we devise a test which tests the relevant aspects of bark-spider without relying on the details in brooks?  Should we  be mocking the simulator in bark-spider?

**Comment by abingham** [comment]
That's a good question. On the one hand, we're not terribly concerned about
the values themselves, just the fidelity of the pipe from brooks to the
browser. On the other, an end-to-end test is the best way to test the pipe.
Maybe what we need to do is just run the simulation directly (i.e. without
the bark-spider machinery in the way) and compare those results with those
retrieved through bark-spider.

**Comment by rob-smallshire** [comment]
That's an excellent idea. It will allow brooks to evolve and improve without us constantly having to update this test, while retaining the benefits of end-to-end testing.  I'm not generally in favour of mocking unless it's absolutely necessary.

**Comment by rob-smallshire** [comment]
Closed by 6cc06ecd
